jupyter-org-client.el

```diff
@@ -416,6 +416,357 @@ the `syntax-table' will be set to that of the REPL buffers."
+  "Kill the block and its results."
+  (interactive)
+  (let ((src (org-element-context))
+        (result-start (org-babel-where-is-src-block-result))
+        end)
+    (if result-start
+        (save-excursion
+          (goto-char result-start)
+          (setq end (org-element-property :end src)))
+      (setq end (org-element-property :end src)))
+    (kill-region
+     (org-element-property :begin src)
+     end)))
+
+;;;###autoload
+(defun jupyter-org-copy-block-and-results ()
+  "Copy the src block at the current point and its results."
+  (interactive)
+  (let ((src (org-element-context))
+        (result-start (org-babel-where-is-src-block-result))
+        end)
+    (if result-start
+        (save-excursion
+          (goto-char result-start)
+          (setq end (org-babel-result-end)))
+      (setq end (org-element-property :end src)))
+    (kill-new
+     (buffer-substring
+      (org-element-property :begin src)
+      end))))
+
+;;;###autoload
+(defun jupyter-org-clone-block (&optional below)
+  "Clone the block."
+  (interactive "P")
+  (let* ((src (org-element-context))
+         (code (org-element-property :value src)))
+    (unless (eq (org-element-type src) 'src-block)
+      (error "Not in a source block"))
+    (jupyter-org-insert-src-block (not below))
+    (delete-char 1)
+    (insert code)
+    ;; jump back to start of new block
+    (org-babel-previous-src-block)
+    (org-babel-next-src-block)))
+
+;;;###autoload
+(defun jupyter-org-merge-blocks ()
+  "Merge the current block with the next block."
+  (interactive)
+  (let ((current-src-block (org-element-context))
+        (next-src-block (save-excursion
+                          (org-babel-next-src-block)
+                          (org-element-context))))
+    (unless (eq (org-element-type current-src-block) 'src-block)
+      (error "Not in a source block"))
+    (let ((merged-code (concat (org-element-property :value current-src-block)
+                               (org-element-property :value next-src-block)))
+          (lang (org-element-property :language current-src-block))
+          (switches (or (org-element-property :switches current-src-block) ""))
+          (parameters (or (org-element-property :parameters current-src-block) "")))
+      ;; Remove source blocks
+      (mapc (lambda (src-block)
+              (goto-char (org-element-property :begin src-block))
+              (org-babel-remove-result)
+              (setf (buffer-substring (org-element-property :begin src-block)
+                                      (org-element-property :end src-block))
+                    ""))
+            (list next-src-block current-src-block))
+      ;; Now create the merged block, point is where the current block was
+      (insert (format "#+BEGIN_SRC %s %s %s\n%s#+END_SRC\n\n"
+                      lang switches parameters merged-code))
+      (forward-line -3)
+      (end-of-line))))
+
+;;;###autoload
+(defun jupyter-org-move-src-block (&optional below)
```

Comment:
Can `org-element-swap-A-B` used to do the work?

Reply:
The problem is that we need to also move the results of a block (if any).
Using `org-element-swap-A-B` would rather complicate things.